@amaltaro could you please look at the Final version of this PR.  And let me mention about the unit tests, before you started asking where are they :) Last night I spent some time trying to setup and develop those but because of MongoDB dependencies this task turned into a non trivial one. For testing any of the methods inside the `MSOutput` Class, I need an instance of it, but then we need the MongoDb to be up and running in the container for the unit tests - which is not the case for the moment. And the whole thing is breaking with [1]. And even more, I have no idea how would this behave when the Unit test is about to be run by Jenkins. So I did not push whatever I have so far because it will simply continue breaking. 





[1]
```
[dmwm@a4e07b4a650f Unified_t]$ nosetests MSOutput_t.py
/home/dmwm/unittestdeploy/wmagent/1.3.6.pre4/sw/slc7_amd64_gcc630/external/py2-cryptography/2.4.2-comp2/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py:163: CryptographyDeprecationWarning: OpenSSL version 1.0.1 is no longer supported by the OpenSSL project, please upgrade. A future version of cryptography will drop support for it.
  utils.CryptographyDeprecationWarning
E
======================================================================
ERROR: Test the getCampaignConfig method
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/MicroService_t/Unified_t/MSOutput_t.py", line 47, in setUp
    self.ms = MSOutput(self.msConfig, 'MSOutputConsumer')
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/MicroService/Unified/MSOutput.py", line 134, in __init__
    self.msOutDB = MongoDB(**msOutDBConfig).msOutDB
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Database/MongoDB.py", line 47, in __init__
    raise ex
ServerSelectionTimeoutError: localhost:8230: [Errno 111] Connection refused
-------------------- >> begin captured logging << --------------------
root: DEBUG: Reading mocked SiteDB data from the file /home/dmwm/wmcore_unittest/WMCore/test/python/../data/Mock/SiteDBMockData.json 
root: INFO: Configuration including default values:
...

root: DEBUG: Initializing CRIC with url: https://cms-cric.cern.ch/
root: ERROR: Could not connect to MongoDB server: mongodb://localhost
localhost:8230: [Errno 111] Connection refusedGiving up Now.
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 30.127s

FAILED (errors=1)
```

--- a/src/python/WMCore/MicroService/Unified/MSOutput.py
+++ b/src/python/WMCore/MicroService/Unified/MSOutput.py
@@ -24,6 +24,7 @@ from WMCore.MicroService.DataStructs.DefaultStructs import OUTPUT_CONSUMER_REPOR
 from WMCore.MicroService.Unified.MSCore import MSCore
 from WMCore.Services.DDM.DDM import DDM, DDMReqTemplate
 from WMCore.Services.CRIC.CRIC import CRIC
+from WMCore.Services.Rucio.Rucio import Rucio
 from Utils.EmailAlert import EmailAlert
 from Utils.Pipeline import Pipeline, Functor
 from WMCore.Database.MongoDB import MongoDB
@@ -51,6 +52,18 @@ class EmptyResultError(MSOutputException):
         else:
             self.myMessage = "EmptyResultError."
         super(EmptyResultError, self).__init__(self.myMessage)
+
+
+class UnsupportedError(MSOutputException):
+    """
+    A MSOutputException signalling an unsupported mode for a function or method.
+    """
+    def __init__(self, message=None):
+        if message:
+            self.myMessage = "UnsupportedError: %s"
+        else:
+            self.myMessage = "UnsupportedError."
+        super(UnsupportedError, self).__init__(self.myMessage)
 
 
 class MSOutput(MSCore):
@@ -109,9 +122,13 @@ class MSOutput(MSCore):
         self.msOutDB = MongoDB(**msOutDBConfig).msOutDB
         self.msOutRelValColl = self.msOutDB['msOutRelValColl']
         self.msOutNonRelValColl = self.msOutDB['msOutNonRelValColl']
-        self.ddm = DDM(url=self.msConfig['ddmUrl'],
-                       logger=self.logger,
-                       enableDataPlacement=self.msConfig['enableDataPlacement'])
+        if self.msConfig['defaultDataManSys'] == 'DDM':
+            self.ddm = DDM(url=self.msConfig['ddmUrl'],
+                           logger=self.logger,
+                           enableDataPlacement=self.msConfig['enableDataPlacement'])
+        elif self.msConfig['defaultDataManSys'] == 'Rucio':
+            self.rucio = Rucio(self.msConfig['rucioAccount'],
+                               configDict={"logger": self.logger})
 
     @retry(tries=3, delay=2, jitter=2)
     def updateCaches(self):
@@ -269,7 +286,7 @@ class MSOutput(MSCore):
         # NOTE:
         #    Here is just an example construction of the function. None of the
         #    data structures used to visualise it is correct. To Be Updated
- 
+
         if self.msConfig['defaultDataManSys'] == 'DDM':
             # NOTE:
             #    We always aggregate per workflow here (regardless of enableAggSubscr)
@@ -320,8 +337,7 @@ class MSOutput(MSCore):
 
                 try:
                     # In the message bellow we may want to put the list of datasets too
-                    msg = "Making transfer subscriptions for %s"
-                    self.logger.info(msg, workflow['RequestName'])
+                    self.logger.info("Making transfer subscriptions for %s", workflow['RequestName'])
 
                     if ddmReqList:
                         ddmResultList = self.ddm.makeAggRequests(ddmReqList, aggKey='item')
@@ -330,7 +346,8 @@ class MSOutput(MSCore):
                         #    Nothing else to be done here. We mark the document as
                         #    done so we do not iterate through it multiple times
                         msg = "Skip submissions for %s. Either all data Tiers were "
-                        msg += "excluded or there were no Output Datasets at all."
+                        msg += "excluded or there were no Output Datasets at all. "
+                        msg += "Marking this workflow as `done`."
                         self.logger.warning(msg, workflow['RequestName'])
                         self.docKeyUpdate(workflow, transferStatus='done')
                         return workflow
@@ -388,17 +405,112 @@ class MSOutput(MSCore):
                     #    reconstructing and returning the same type of object
                     #    as the one that have been passed to the current call.
                     pass
+                # FIXME:
+                msg = "Not yet implemented mode with workflows of type %s!\n" % type(workflow)
+                msg += "Skipping this call"
+                self.logger.error(msg)
+                raise NotImplementedError
             else:
                 msg = "Unsupported type %s for workflows!\n" % type(workflow)
                 msg += "Skipping this call"
                 self.logger.error(msg)
+                raise UnsupportedError
 
         elif self.msConfig['defaultDataManSys'] == 'PhEDEx':
             pass
 
         elif self.msConfig['defaultDataManSys'] == 'Rucio':
-            pass
+            if isinstance(workflow, MSOutputTemplate):
+                self.logger.info("Making transfer subscriptions for %s", workflow['RequestName'])
 
+                rucioResultList = []
+                if workflow['destinationOutputMap']:
+                    for dMap in workflow['destinationOutputMap']:
+                        try:
+                            copies = workflow['numberOfCopies'] if workflow['numberOfCopies'] else 1
+                            # NOTE:
+                            #    Once we get rid of DDM this rseExpression generation
+                            #    should go in the Producer thread
+                            if workflow['isRelVal']:
+                                if dMap['destination']:
+                                    rseUnion = '('+'|'.join(dMap['destination'])+')'
+                                else:
+                                    # NOTE:
+                                    #    If we get to here it is most probably because the destination
+                                    #    in the destinationOutputMap is empty (due to an
+                                    #    'ALCARECO' dataset from a Relval workflow or similar).
+                                    #    Since this is expected to happen a lot, we'd better just
+                                    #    log a warning and continue
+                                    msg = "No destination provided. Avoid creating transfer subscription for "
+                                    msg += "Workflow: %s : Dataset Names: %s"
+                                    self.logger.warning(msg, workflow['RequestName'], dMap['datasets'])
+                                    continue
+                                rseExpression = rseUnion + '&cms_type=real&rse_type=DISK'
+                                # NOTE:
+                                #    The above rseExpression should resolve to something similar to:
+                                #    (T2_CH_CERN|T1_US_FNAL_Disk)&cms_type=real&rse_type=DISK
+                                #    where the first part is a Union of all destination sites and
+                                #    the second part is a general constraint for those to be real
+                                #    entries but not `Test` or `Temp` and we also target only sites
+                                #    marked as `Disk`
+                            else:
+                                rseExpression = '(tier=2|tier=1)&cms_type=real&rse_type=DISK'
+                                # NOTE:
+                                #    The above rseExpression should target all T1_*_Disk and T2_*
+                                #    sites, where the first part is a Union of those Tiers and
+                                #    the second part is a general constraint for those to be real
+                                #    entries but not `Test` or `Temp` and we also target only sites
+                                #    marked as `Disk`
+
+                            if self.msConfig['enableDataPlacement']:
+                                rucioResultList.append(self.rucio.createReplicationRule(dMap['datasets'],
+                                                                                        rseExpression,
+                                                                                        copies=copies))
+                            else:
+                                msg = "DRY-RUN:: The effective Rucio submission would look like: \n"
+                                msg += "account: %s \n"
+                                msg += "dids: %s \n"
+                                msg += "rseExpression: %s\n"
+                                msg += "copies: %s\n"
+                                self.logger.warning(msg,
+                                                    self.msConfig['rucioAccount'],
+                                                    pformat(dMap['datasets']),
+                                                    rseExpression,
+                                                    copies)
+                        except Exception as ex:
+                            msg = "Could not make transfer subscription for Workflow: %s\n:%s"
+                            self.logger.exception(msg, workflow['RequestName'], str(ex))
+                            return workflow
+                else:
+                    # NOTE:
+                    #    Nothing else to be done here. We mark the document as
+                    #    done so we do not iterate through it multiple times
+                    msg = "Skip submissions for %s. Either all data Tiers were "
+                    msg += "excluded or there were no Output Datasets at all. "
+                    msg += "Marking this workflow as `done`."
+                    self.logger.warning(msg, workflow['RequestName'])
+                    self.docKeyUpdate(workflow, transferStatus='done')
+                    return workflow
+
+                transferIDs = rucioResultList
+                if self.msConfig['enableDataPlacement']:
+                    self.docKeyUpdate(workflow,
+                                      transferStatus='done',
+                                      transferIDs=transferIDs)
+                return workflow
+
+            elif isinstance(workflow, (list, set, CommandCursor)):
+                # FIXME:
+                msg = "Not yet implemented mode with workflows of type %s!\n" % type(workflow)
+                msg += "Skipping this call"
+                self.logger.error(msg)
+                raise NotImplementedError
+
+            else:
+                msg = "Unsupported type %s for workflows!\n" % type(workflow)
+                msg += "Skipping this call"
+                self.logger.error(msg)
+                raise UnsupportedError
         # NOTE:
         #    if we are about to implement this through a pipeline we MUST not
         #    return the result here but the WHOLE document with updated fields
@@ -542,7 +654,7 @@ class MSOutput(MSCore):
 
         wfCounterTotal = sum(wfCounters.values())
         return wfCounterTotal
- 
+
     def msOutputProducer(self, requestRecords):
         """
         A top level function to drive the upload of all the documents to MongoDB
@@ -712,7 +824,7 @@ class MSOutput(MSCore):
                 #    Once we migrate to Rucio we should change those defaults to
                 #    whatever is the format in Rucio (eg. referring a subscription
                 #    rule like: "store it at a good site" or "Store in the USA" etc.)
-                destination.add('T1_*')
+                destination.add('T1_*_Disk')
                 destination.add('T2_*')
 
             dMap = {'datasets': [dataset],
@@ -740,7 +852,7 @@ class MSOutput(MSCore):
                         for i in dMap['datasets']:
                             aggMap['datasets'].append(i)
                     found = True
-                    del(dMap)
+                    del dMap
                     break
             if not found:
                 aggDstMap.append(dMap)

--- a/src/python/WMCore/Services/Rucio/Rucio.py
+++ b/src/python/WMCore/Services/Rucio/Rucio.py
@@ -9,8 +9,10 @@ from __future__ import division, print_function, absolute_import
 
 import logging
 from copy import deepcopy
+from pprint import pformat
+
 from rucio.client import Client
-from rucio.common.exception import (AccountNotFound, DataIdentifierNotFound, AccessDenied,
+from rucio.common.exception import (AccountNotFound, DataIdentifierNotFound, AccessDenied, DuplicateRule,
                                     DataIdentifierAlreadyExists, DuplicateContent,
                                     UnsupportedOperation, FileAlreadyExists, RuleNotFound)
 from WMCore.WMException import WMException
@@ -478,9 +480,82 @@ class Rucio(object):
         except AccessDenied as ex:
             msg = "AccessDenied creating DID replication rule. Error: %s" % str(ex)
             raise WMRucioException(msg)
+        except DuplicateRule as ex:
+            # NOTE:
+            #    The unique constraint is per did and it is checked against the tuple:
+            #    rucioAccount + did = (name, scope) + rseExpression
+            # NOTE:
+            #    This exception will be thrown by Rucio even if a single Did has
+            #    a duplicate rule. In this case all the rest of the Dids will be
+            #    ignored, which in general should be addressed by Rucio. But since
+            #    it is not, we should break the list of Dids and proceed one by one
+            # NOTE:
+            #    This thing here may be slow, because it will wait for Rucio to
+            #    return the history of rules per every Did, but no shorter path exists
+            msg = "A duplicate rule for: \naccount: %s \ndids: %s \nrseExpression: %s.\n"
+            self.logger.info(msg,
+                             kwargs['account'],
+                             pformat(dids),
+                             rseExpression)
+
+            ruleIds = []
+            didsDup = []
+            for did in dids:
+                try:
+                    response = self.cli.add_replication_rule([did], copies, rseExpression, **kwargs)
+                    for ruleId in response:
+                        ruleIds.append(ruleId)
+                    self.logger.debug("Per did ruleIds: %s", ruleIds)
+                except DuplicateRule:
+                    didsDup.append(did)
+
+            ruleHistory = self.listRuleHistory(didsDup)
+            self.logger.debug("Rule History: %s\n", pformat(ruleHistory))
+
+            for did in ruleHistory:
+                for didHist in did['did_hist']:
+                    ruleIds.append(didHist['rule_id'])
+            ruleIds = list(set(ruleIds))
+            self.logger.debug("ruleIds: %s\n", ruleIds)
+            return ruleIds
         except Exception as ex:
             self.logger.error("Exception creating rule replica for data: %s. Error: %s", names, str(ex))
         return response
+
+    def listRuleHistory(self, dids):
+        """
+        _listRuleHistory_
+
+        A function to return a list of historical records of replication rules
+        per did.
+        :param dids: a list of dids of the form {'name': '...', 'scope: '...'}
+
+        The returned structure looks something like:
+        [{'did': {'name': 'DidName',
+                  'scope': 'cms'},
+          'did_hist': [{u'account': u'wma_test',
+                        u'created_at': datetime.datetime(2020, 6, 30, 1, 34, 51),
+                        u'locks_ok_cnt': 9,
+                        u'locks_replicating_cnt': 0,
+                        u'locks_stuck_cnt': 0,
+                        u'rse_expression': u'(tier=2|tier=1)&cms_type=real&rse_type=DISK',
+                        u'rule_id': u'1f0ab297e4b54e1abf7c086ac012b9e9',
+                        u'state': u'OK',
+                        u'updated_at': datetime.datetime(2020, 6, 30, 1, 34, 51)}]},
+         ...
+         {'did': {},
+          'did_hist': []}]
+        """
+        fullHistory = []
+        for did in dids:
+            didHistory = {}
+            didHistory['did'] = did
+            didHistory['did_hist'] = []
+            # check the full history of the current did
+            for hist in self.cli.list_replication_rule_full_history(did['scope'], did['name']):
+                didHistory['did_hist'].append(hist)
+            fullHistory.append(didHistory)
+        return fullHistory
 
     def listContent(self, name, scope='cms'):
         """

